arm64: dts: qcom: msm8916-samsung-fortuna/rossa: Upstreaming

wonderfulShrineMaidenOfParadise opened the pull request:
Discussion thread before upstreaming.

## Todo
- [x] Rename the device trees like downstream does (fortuna-*)
- [x] Or use the codenames in `build.prop`
- [ ] ~~Drop~~ Redo `samsung,fortuna`
- [ ] ~~Squash commits (Sensors, NFC...)~~
- [ ] Enable NFC for grandprimelte if there is anyone interested
- [ ] Add coreprimeltespr if there is anyone interested
- [ ] Add cprimeltetmo if there is anyone interested
- [ ] Add fortuna3gdtv if there is anyone interested. Cc @TotallyNotChloe 
- [ ] Add fortunaltedx if there is anyone interested. Cc @BaranSenkul
- [ ] Add fortunalteub if there is anyone interested. Cc @pachof

|Model|Current DTS|Codename|Downstream DTS|
|-----|-----------|--------|--------------|
|SM-G530BT|None|fortuna3gdtv|fortunadtv|
|SM-G530F|None|fortunaltedx|fortuna-sea|
|SM-G530FZ|[grandprimelte](https://github.com/torvalds/linux/blob/master/arch/arm64/boot/dts/qcom/msm8916-samsung-grandprimelte.dts)|grandprimelte|[fortuna-eur](https://github.com/msm8916-mainline/linux-downstream/blob/SM-G530FZ/arch/arm/boot/dts/samsung/msm8916/msm8916-sec-fortuna-eur-r02.dts)|
|SM-G530H|[fortuna3g](https://github.com/torvalds/linux/blob/master/arch/arm64/boot/dts/qcom/msm8216-samsung-fortuna3g.dts)|fortuna(ve)3g|[fortuna-r11](https://github.com/msm8916-mainline/linux-downstream/blob/SM-G530FZ/arch/arm/boot/dts/samsung/msm8916/msm8916-sec-fortuna-r11.dts)/[fortuna-ve3g](https://github.com/msm8916-mainline/linux-downstream/blob/SM-G530FZ/arch/arm/boot/dts/samsung/msm8916/msm8916-sec-fortuna-ve3g-r07.dts)|
|SM-G530M|None|fortunalteub|[fortuna-r11](https://github.com/msm8916-mainline/linux-downstream/blob/SM-G530FZ/arch/arm/boot/dts/samsung/msm8916/msm8916-sec-fortuna-r11.dts)|
|SM-G530MU/Y|None -> [fortunaltezt](https://github.com/msm8916-mainline/linux/blob/msm8916/6.6/arch/arm64/boot/dts/qcom/msm8916-samsung-fortunaltezt.dts)|fortunaltezt|[fortuna-r11](https://github.com/msm8916-mainline/linux-downstream/blob/SM-G530Y/arch/arm/boot/dts/samsung/msm8916-sec-fortuna-r11.dts) (another variant)|
|SM-G530W|[gprimeltecan](https://github.com/torvalds/linux/blob/master/arch/arm64/boot/dts/qcom/msm8916-samsung-gprimeltecan.dts)|gprimeltecan|[fortuna-tmo](https://github.com/msm8916-mainline/linux-downstream/blob/SM-G530W/arch/arm/boot/dts/samsung/msm8916/msm8916-sec-fortuna-tmo-r02.dts)|
|SM-G360G|[rossa](https://github.com/torvalds/linux/blob/master/arch/arm64/boot/dts/qcom/msm8916-samsung-rossa.dts)|rossaltexsa|[rossa-eur](https://github.com/msm8916-mainline/linux-downstream/blob/SM-G360G/arch/arm/boot/dts/samsung/msm8916/msm8916-sec-rossa-eur-r01.dts)|
|SM-G360P|None|coreprimeltespr|[rossaspr](https://github.com/msm8916-mainline/linux-downstream/blob/SM-G360P/arch/arm/boot/dts/samsung/msm8916/msm8916-sec-rossaspr-r03.dts)|
|SM-G360T|None|cprimeltetmo|[rossatmo](https://github.com/msm8916-mainline/linux-downstream/blob/SM-G530T/arch/arm/boot/dts/samsung/msm8916/msm8916-sec-rossatmo-r00.dts)|
|SM-G357FZ|None -> [heatqlte](https://github.com/msm8916-mainline/linux/blob/msm8916/6.6/arch/arm64/boot/dts/qcom/msm8916-samsung-heatqlte.dts)|heatqlte|[heat](https://github.com/msm8916-mainline/linux-downstream/blob/SM-G357FZ/arch/arm/boot/dts/samsung/msm8916-sec-heat-r02.dts)|

## Current conclusions
- Resort commits
- Apply new dt-schema
- Rename `gprime` into fortuna and `cprime` into `rossa` 
- Keep using codenames in `build.prop`
- Move `fortuna3g` to `msm8216`
- Add `grandprimelte`
- Replace gpio-vibrator with regulator-haptic
- Move `tz-apps@85a00000` to `msm8916-samsung-fortuna-common.dtsi`
- `interrupts-extended` will be used instead of `interrupt-parent` and `interrupts`, as preferred in upstream.
- The following will be upstreamed:
  - [x] `fortuna3g` Cc @SiddharthManthan 
  - [ ] ~~`fortunaltezt`~~ (postponed due to tsens issue)
  - [x] `grandprimelte`
  - [x] `gprimeltecan` Cc @buddyjojo @WTechNinja
  - [ ] ~~`heatqlte`~~ (postponed due to tsens issue) Cc @garethppls
  - [x] `rossa`
- To be upstreamed:
  - [x] Initial device trees, sound and modem
  https://lore.kernel.org/r/20240129143147.5058-1-raymondhackley@protonmail.com
  - [x] Fuel gauge
  https://lore.kernel.org/r/20240216124639.24689-1-raymondhackley@protonmail.com
  - [x] Touchscreen on fortuna
  https://lore.kernel.org/r/20240404121703.17086-2-raymondhackley@protonmail.com
  - [x] PWM backlight
  https://lore.kernel.org/r/20240404121703.17086-3-raymondhackley@protonmail.com
  - [x] Micro USB connector
  https://lore.kernel.org/r/20240424144922.28189-1-raymondhackley@protonmail.com
  - [x] Accelerometer / Magnetometer
  https://lore.kernel.org/r/20240405120803.20754-1-raymondhackley@protonmail.com
  - [x] NFC
  https://lore.kernel.org/r/20240601115321.25314-2-raymondhackley@protonmail.com
  https://lore.kernel.org/r/20240601115321.25314-3-raymondhackley@protonmail.com
  - [x] PMIC/Charger
  https://lore.kernel.org/r/20240601115321.25314-4-raymondhackley@protonmail.com
  - [x] Touchscreen on rossa
  https://lore.kernel.org/r/20240723131441.1764-1-raymondhackley@protonmail.com
  - [x] Zinitix touch keys on fortuna
  https://lore.kernel.org/r/20240724143230.3804-1-raymondhackley@protonmail.com

buddyjojo committed:
FIXME: arm64: dts: qcom: msm8916-samsung-fortuna: Add audio jack detection

Signed-off-by: Joe Mason <buddyjojo06@outlook.com>
[Raymond: Split from the audio patch]
Signed-off-by: Raymond Hackley <raymondhackley@protonmail.com>

bolilingmeng89006 committed:
arm64: dts: qcom: msm8916-samsung-fortuna: Add PN547 NFC (v3)

Grand Prime SM-G530Y (fortunaltezt) has a NXP PN547, which is supported
by the nxp-nci-i2c driver in mainline. It seems to detect NFC tags
using "nfctool" just fine, although more testing is difficult given
there seem to be very few useful applications making use of the
Linux NFC subsystem. :(

[Stephan: Move i2c-nfc to common dtsi to share it with other variants]
Signed-off-by: Stephan Gerhold <stephan@gerhold.net>
v2: Reword and add label pn547_nfc
v3: Use interrupts-extended
Signed-off-by: "Lin, Meng-Bo" <linmengbo0689@protonmail.com>

bolilingmeng89006 committed:
ARM: dts: qcom-msm8916-samsung-fortunaltezt/heatqlte: Include dts from arm64

After adding all necessary support for MSM8916 SMP/cpuidle without PSCI
on ARM32, build fortunaltezt and heatqlte device trees from the arm64
tree together with the ARM32 include to allow booting this device on ARM32.

The approach to include device tree files from other architectures is
inspired from e.g. the Raspberry Pi (bcm2711-rpi-4-b.dts) where this is
used to build the device tree for both ARM32 and ARM64.

[Stephan: Add workaround to TSENS problem]
Signed-off-by: Stephan Gerhold <stephan@gerhold.net>
[Gareth: Add heatqlte]
Signed-off-by: Gareth Peoples <mail@gpeopl.es>
Signed-off-by: "Lin, Meng-Bo" <linmengbo0689@protonmail.com>

bolilingmeng89006 committed:
ARM: dts: qcom-msm8916-samsung-fortunaltezt/heatqlte: Add qcom,srot-locked to workaround TSENS problem

Signed-off-by: Stephan Gerhold <stephan@gerhold.net>
Signed-off-by: Gareth Peoples <mail@gpeopl.es>
Signed-off-by: "Lin, Meng-Bo" <linmengbo0689@protonmail.com>

garethppls committed:
arm64: dts: qcom: msm8916-samsung-heatqlte: Add display panel

Signed-off-by: Gareth Peoples <mail@gpeopl.es>
[Raymond: Split from the init commit]
Signed-off-by: Raymond Hackley <raymondhackley@protonmail.com>

garethppls committed:
arm64: dts: qcom: msm8916-samsung-heatqlte: Add initial device tree

Samsung Galaxy Ace 4 is a phone based on MSM8916. It is similar
to the other Samsung devices based on MSM8916 with only a few minor
differences.

Ace 4 and Core Prime are similar, with some differences in panel and
touchscreen. The common parts are shared in
msm8916-samsung-rossa-common.dtsi to reduce duplication.

Unfortunately, SM-G357FZ was released with outdated 32-bit only
firmware and never received any update from Samsung. Since the 32-bit
TrustZone firmware is signed there seems to be no way currently to
actually boot this device tree on arm64 Linux at the moment.

However, it is possible to use this device tree by compiling an ARM32
kernel instead. The device tree can be easily built on ARM32 with
an #include and it works really well there. To avoid confusion for others
it is still better to add this device tree on arm64. Otherwise it's easy
to forget to update this one when making some changes that affect all
MSM8916 devices.

Maybe someone finds a way to boot ARM64 Linux on this device at some
point.
In this case we expect that this device tree can be simply used as-is.

Signed-off-by: Gareth Peoples <mail@gpeopl.es>
[Raymond: Use msm8916-samsung-rossa-common.dtsi and reword the commit]
Signed-off-by: Raymond Hackley <raymondhackley@protonmail.com>

wonderfulShrineMaidenOfParadise committed:
arm64: dts: qcom: msm8916-samsung-rossa: Add display panel

Signed-off-by: Raymond Hackley <raymondhackley@protonmail.com>

wonderfulShrineMaidenOfParadise commented on arch/arm64/boot/dts/qcom/msm8916-samsung-heat.dts:
```diff
@@ -0,0 +1,11 @@
+
+/ {
+	model = "Samsung Galaxy Ace 4 (SM-G357FZ)";
+	compatible = "samsung,heatqlte", "qcom,msm8916";
```
There are various of `heat`s. `heatqlte` is the only known qcom variant.

wonderfulShrineMaidenOfParadise commented on arch/arm64/boot/dts/qcom/msm8916-samsung-rossa-eur.dts:
```diff
@@ -0,0 +1,11 @@
+#include "msm8916-samsung-rossa.dtsi"
+
+/ {
+	model = "Samsung Galaxy Core Prime LTE (SM-G360G)";
```
```suggestion
	model = "Samsung Galaxy Core Prime LTE";
```

Could be also compatible with SM-G360F. `LTE` is kept here since all MSM8916 Core Prime has LTE support.

stephan-gh replied:
You could add `Core Prime LTE (rossa)` or `Core Prime LTE (MSM8916)` if you want to make it clear that the MSM8916 variants are meant.

stephan-gh replied:
The main purpose of `model` is to allow identifying the device tree in a readable way, it doesn't have to include all supported model numbers.

wonderfulShrineMaidenOfParadise commented on arch/arm64/boot/dts/qcom/msm8216-samsung-fortuna3g.dts:
```diff
@@ -2,11 +2,11 @@
 / {
 	model = "Samsung Galaxy Grand Prime (SM-G530H)";
-	compatible = "samsung,fortuna3g", "samsung,gprime", "qcom,msm8916";
+	compatible = "samsung,fortuna3g", "samsung,fortuna", "qcom,msm8916";
```
We probably don't need this?

stephan-gh replied:
I don't think there is any use case for a fallback compatible in Linux atm, you'd just risk having this DTB selected on the wrong device.

wonderfulShrineMaidenOfParadise replied:
We found the usage for Sxmo profiles, so add it as `samsung,fortuna`.

wonderfulShrineMaidenOfParadise commented:
- Rebased on 6.5-rc4
- Applied latest mailing list patch set
- Put panel and modem commits after the other patches

stephan-gh commented on arch/arm/boot/dts/qcom/qcom-msm8916-samsung-fortunaltezt.dts:
```diff
@@ -3,6 +3,6 @@
-	/* The device crashes when accessing the SROT region for some reason */
-	qcom,srot-locked;
+	/* FIXME: The device crashes when accessing the SROT region for some reason */
+	status = "disabled";
```
Can you add a commit on top that keeps the srot-locked for msm8916-mainline?

wonderfulShrineMaidenOfParadise replied:
It was accidentally dropped. I will bring it back.

wonderfulShrineMaidenOfParadise commented on arch/arm64/boot/dts/qcom/msm8916-samsung-fortuna-common.dtsi:
```diff
@@ -290,7 +309,7 @@
 
 	panel: panel@0 {
-		compatible = "samsung,gprime-panel";
+		compatible = "samsung,fortuna-panel";
```
msm8916-mainline/lk2nd#282

wonderfulShrineMaidenOfParadise commented on arch/arm64/boot/dts/qcom/msm8916-samsung-rossa.dts:
```diff
@@ -36,5 +36,5 @@
 
 &panel {
-	compatible = "samsung,cprime-panel";
+	compatible = "samsung,rossa-panel";
```
msm8916-mainline/lk2nd#282

wonderfulShrineMaidenOfParadise reviewed (commented):
Notes about rename:

wonderfulShrineMaidenOfParadise commented:
For v3 upstream patch, I would add the following changes:
- Temporarily drop fortunaltezt and heatqlte. I don't know how to reply to TSENS issues and it's not in hurry.
- Add sound and modem to initial device trees.
- Minor fixes for signed-off-by tags.

bolilingmeng89006 committed:
arm64: dts: qcom: msm8916-samsung-fortunaltezt: Add initial device tree

Samsung Galaxy Grand Prime (SM-G530Y) is a smartphone using the MSM8916
SoC released in 2014.

The Grand Prime variants are very similar, with some differences in
accelerometer and NFC. The common parts are shared in
msm8916-samsung-fortuna-common.dtsi to reduce duplication.

Unfortunately, SM-G530Y was released with outdated 32-bit only
firmware and never received any update from Samsung. Since the 32-bit
TrustZone firmware is signed there seems to be no way currently to
actually boot this device tree on arm64 Linux at the moment.

However, it is possible to use this device tree by compiling an ARM32
kernel instead. The device tree can be easily built on ARM32 with
an #include and it works really well there. To avoid confusion for others
it is still better to add this device tree on arm64. Otherwise it's easy
to forget to update this one when making some changes that affect all
MSM8916 devices.

Maybe someone finds a way to boot ARM64 Linux on this device at some
point.
In this case we expect that this device tree can be simply used as-is.

Signed-off-by: "Lin, Meng-Bo" <linmengbo0689@protonmail.com>

cristianc commented:
Patches have been accepted/merged upstream.

stephan-gh commented:
I think only the dt-bindings patch was applied because the dts one was missing a file (see https://lore.kernel.org/linux-arm-msm/th4btdwcsvvfbipshjeplvd7kuyd6qxuv5odbksydc23zfls47@awgmdy3u3ck5/). @wonderfulShrineMaidenOfParadise can you resend that patch with the missing file?

wonderfulShrineMaidenOfParadise committed:
arm64: dts: qcom: msm8916-samsung-fortuna: Add display panel

Signed-off-by: Raymond Hackley <raymondhackley@protonmail.com>

wonderfulShrineMaidenOfParadise commented:
v4 Link: https://lore.kernel.org/all/20240129143147.5058-1-raymondhackley@protonmail.com/

wonderfulShrineMaidenOfParadise commented on arch/arm64/boot/dts/qcom/msm8916-samsung-grandprimelte.dts:
```diff
@@ -0,0 +1,30 @@
+	chassis-type = "handset";
+};
+
+&battery {
+	charge-term-current-microamp = <200000>;
+	constant-charge-current-max-microamp = <1000000>;
+	constant-charge-voltage-max-microvolt = <4350000>;
+};
```
Just a note. I forgot to add these and both rt5033-charger and rt5033-battery (fuel gauge) failed.
Cc: @Jakko3

Jakko3 replied:
You mean: When you don't add these lines, then the rt5033-charger and rt5033-battery fail?

The rt5033-charger is supposed to fail. It needs to know the battery data to work. If some of the battery data is missing, the driver "exits" on purpose. The charger chip then charges without the driver on the default values programmed into the chip (rather low current, unfortunately up to 4.4 V).

The rt5033-battery (fuel gauge) is not supposed to fail. It's supposed to go to charging status POWER_SUPPLY_STATUS_UNKNOWN because the rt5033-battery (fuel gauge) is not able to detect the charging status itself without the charger driver. At some point in the past, the battery icon in e.g. Phosh in that case showed a question mark for some time. By now, UPower should guess this as "discharging" (no matter wheter the battery is charging or not, as this cannot be detected on rt5033 without the charger driver running).

wonderfulShrineMaidenOfParadise replied:
@Jakko3 

> The rt5033-battery (fuel gauge) is not supposed to fail.

```
cat: '/sys/class/power_supply/rt5033-*': No such file or directory
$ lsmod|grep rt5033
rt5033_regulator       12288  0 
rt5033_battery         12288  0 
rt5033_charger         12288  0 
rt5033                 12288  0 
mfd_core               12288  1 rt5033
```

```
[    1.509005] rt5033 5-0034: Device found (rev. 5)
[    1.511150] rt5033-charger rt5033-charger: DMA mask not set
[    1.514747] rt5033-charger rt5033-charger: Value 'constant-charge-current-max-microamp' out of range
[    1.514758] rt5033-charger rt5033-charger: probe with driver rt5033-charger failed with error -22
```

By dropping the properties, I just reproduced it on 6.9. I am not sure why, but the fuel gauge actually failed along with the charger.

Jakko3 replied:
I checked my testing notes from development phase. I see that the rt5033-battery driver fails when the modules `rt5033`, `rt5033-regulator` & `rt5033-charger` are added to initramfs (not adding the `rt5033-battery` module, though).

I don't know why adding them to initramfs makes the rt5033-battery driver fail if a charging parameter is missing. When they're not added to initramfs, the rt5033-battery driver doesn't fail. (Maybe you can try with adding rt5033-battery to initramfs as well.)

Personally I didn't add them to initramfs because I had no need to. I remember that we talked about adding them to initramfs to have USB OTG keyboard/mouse available at full-disk encryption unlock interface.

Whatsoever, adding the rt5033-charger driver to the devicetree requires to add all five charging parameters.

wonderfulShrineMaidenOfParadise committed:
FROMLIST: arm64: dts: qcom: msm8916/8216-samsung-fortuna: Add samsung,fortuna binding

Add common binding samsung,fortuna used by the following
Samsung Galaxy Grand Prime devices:
- SM-G530FZ (msm8916-samsung-grandprimelte)
- SM-G530H  (msm8216-samsung-fortuna3g)
- SM-G530W  (msm8916-gprimeltecan)

Signed-off-by: Raymond Hackley <raymondhackley@protonmail.com>
Link: https://lore.kernel.org/lkml/20240326141222.200452-2-raymondhackley@protonmail.com

pachof commented:
Hello, I am testing postmarketOS on my Samsung and one of the biggest problems is that the sdcard does not work and there is very little space when installed, if the sdcard worked I would install it in the sdcard and it would have more space

wonderfulShrineMaidenOfParadise commented:
@pachof  please join [#msm8916:postmarketos.org](https://app.element.io/#/room/#msm8916:postmarketos.org). I cannot answer any question on GitHub.

wonderfulShrineMaidenOfParadise committed:
arm64: dts: qcom: msm8916-samsung-fortunaltezt/heatqlte: Add PMIC

The phones listed below have Richtek RT5033 PMIC and charger.
Add them to the device trees.
- Samsung Galaxy Ace 4 (SM-G357FZ)
- Samsung Galaxy Grand Prime (SM-G530Y)

Signed-off-by: Raymond Hackley <raymondhackley@protonmail.com>

bolilingmeng89006 committed:
arm64: dts: qcom: msm8916-samsung-fortunaltezt: Add accelerometer

This device has ST LIS2HH12 accelerometer.
Add support for it.

Signed-off-by: "Lin, Meng-Bo" <linmengbo0689@protonmail.com>

bolilingmeng89006 committed:
arm64: dts: qcom: msm8916-samsung-fortunaltezt: Add display panel

Signed-off-by: "Lin, Meng-Bo" <linmengbo0689@protonmail.com>

wonderfulShrineMaidenOfParadise committed:
UPSTREAM: arm64: dts: qcom: msm8916-samsung-rossa: Add touchscreen

Core Prime uses an Imagis IST3038 touchscreen that is connected to
blsp_i2c5. Add it to the device tree.

Signed-off-by: Raymond Hackley <raymondhackley@protonmail.com>
Reviewed-by: Konrad Dybcio <konrad.dybcio@linaro.org>
Link: https://lore.kernel.org/r/20240723131441.1764-1-raymondhackley@protonmail.com
Signed-off-by: Bjorn Andersson <andersson@kernel.org>

wonderfulShrineMaidenOfParadise committed:
UPSTREAM: arm64: dts: qcom: msm8916-samsung-fortuna: Add touch keys

Touch keys feature on fortuna phones are provided by Zinitix touchscreen.
Add property linux,keycodes to enable touch keys.

Signed-off-by: Raymond Hackley <raymondhackley@protonmail.com>
Reviewed-by: Konrad Dybcio <konrad.dybcio@linaro.org>
Link: https://lore.kernel.org/r/20240724143230.3804-1-raymondhackley@protonmail.com
Signed-off-by: Bjorn Andersson <andersson@kernel.org>
